packages/mobx-react/src/observerClass.ts

```diff
@@ -73,9 +78,26 @@ export function makeClassComponentObserver(
     }
     patch(target, "componentDidMount", function () {
         this[mobxIsUnmounted] = false
-        if (!this.render[mobxAdminProperty]) {
-            // Reaction is re-created automatically during render, but a component can re-mount and skip render #3395.
-            // To re-create the reaction and re-subscribe to relevant observables we have to force an update.
+        recordReactionAsCommitted(this)
+        if (this[reactionTrack]) {
+            this[reactionTrack].mounted = true
+            if (!this.render[mobxAdminProperty] || this[reactionTrack].changedBeforeMount) {
+                // Reaction is re-created automatically during render, but a component can re-mount and skip render #3395.
+                // To re-create the reaction and re-subscribe to relevant observables we have to force an update.
+                Component.prototype.forceUpdate.call(this)
+                this[reactionTrack].changedBeforeMount = false
+            }
+        } else {
```

Comment:
In order for this to happen, you have to unset `this[reactionTrack]` in `componentWillUnmount`, like here
https://github.com/mobxjs/mobx/blob/f0e066f427d573cbef4b92f3310eb069c3aca205/packages/mobx-react-lite/src/useObserver.ts#L103-L104
And also `this[reactionTrack]` must be unset when reaction is disposed in tracking utils, is it the case?